src/storage-types/src/errors.rs

```diff
@@ -444,6 +444,57 @@ pub enum DataflowError {
 
 impl Error for DataflowError {}
 
+mod boxed_str {
+
+    use timely::container::columnation::Region;
+    use timely::container::columnation::StableRegion;
+
+    /// Region allocation for `String` data.
+    ///
+    /// Content bytes are stored in stable contiguous memory locations,
+    /// and then a `String` referencing them is falsified.
+    #[derive(Default)]
+    pub struct BoxStrStack {
+        region: StableRegion<u8>,
+    }
+
+    impl Region for BoxStrStack {
+        type Item = Box<str>;
+        #[inline]
+        fn clear(&mut self) {
+            self.region.clear();
+        }
+        // Removing `(always)` is a 20% performance regression in
+        // the `string10_copy` benchmark.
+        #[inline(always)]
+        unsafe fn copy(&mut self, item: &Box<str>) -> Box<str> {
+            let bytes = self.region.copy_slice(item.as_bytes());
+            Box::from(std::str::from_utf8_unchecked(bytes))
+        }
+        #[inline(always)]
+        fn reserve_items<'a, I>(&mut self, items: I)
+        where
+            Self: 'a,
+            I: Iterator<Item = &'a Self::Item> + Clone,
+        {
+            self.region.reserve(items.map(|x| x.len()).sum());
+        }
+
+        fn reserve_regions<'a, I>(&mut self, regions: I)
+        where
+            Self: 'a,
+            I: Iterator<Item = &'a Self> + Clone,
+        {
+            self.region
+                .reserve(regions.clone().map(|r| r.region.len()).sum());
```

Comment:
Is the `clone` intended here? Looks like this should work without it?

Reply:
Copy/paste leftover! Changed